draft-barnes-mls-appsync.md

```diff
@@ -60,171 +61,225 @@ communicate. Applications based on MLS can integrate their state into this
-application state object.
-
-The content of the `application_states` extension and the `AppSync` proposal are
-structured as follows:
+This document defines a new GroupContextExtensionsDiff proposal. It
+is not a Safe Extension as defined in {{Section 2 of
+!I-D.ietf-mls-extensions}}, because it can modify any GroupContext
+extension, including those defined in {{!RFC9420}}. It is intended as a
+complete replacement for the GroupContextExtensions Proposal.
+The GroupContextExtensionsDiff proposal does not require an UpdatePath,
+and may be sent by an authorized external sender.
 
 ~~~ tls
 enum {
-    irreducible(0),
-    map(1),
-    unorderedList(2),
-    orderedArray(3),
+    remove(0),
+    add(1),
+    replace(2),
+    diff(3),
     (255)
-} StateType;
+} OperationType;
+
+uint8 DiffType;
 
+struct {
+    ExtensionType group_context_extension;
+    OperationType operation;
+    select (operation) {
+        case remove:
+            struct {};
+        case add:
+            opaque extension_data<V>;
+        case replace:
+            opaque extension_data<V>;
+        case diff:
+            opaque diff_data<V>;
+    };
+} ExtensionDiff;
+
+struct {
+    ExtensionDiff group_context_extension_diffs<V>;
+} GroupContextExtensionsDiff;
+~~~
+
+The semantics of OperationType are as follows:
+
+- `remove` means that the GroupContext extension was present, and will be
+completely absent from the GroupContext.
+- `add` means that the GroupContext extension was absent, and will be
+present with the value provided in `extension_data`.
+- `replace` means that the GroupContext extension was present and will
+be completely replaced with the new value in `extension_data`
+- `diff` means that the GroupContext extension will be modified according
+to the diff format defined by that extension. (Two concrete diff formats are
+defined in {{diff-formats}} which extension designers are free to use or
+not use.)
+
+
+## Processing rules
+
+A single GroupContextExtensionsDiff proposal can contain changes to multiple
+extensions. Likewise a single Commit can include multiple
+GroupContextExtensionsDiff proposals, however a single Commit MUST NOT
+include multiple GroupContextExtensionsDiff proposals which refer to the
+same ExtensionType. A GroupContextExtensions proposal MUST NOT appear in a
+commit with any GroupContextExtensionsDiff proposal.
+
+The motivation for allowing multiple extensions in a single
+GroupContextExtensionsDiff proposal is to allow a client to modify multiple
+extensions in a single "transaction". A common example of this is to add a
+GroupContext extension while also adding that extension to the
+`required_capabilities` GroupContext extension.
+
+A proposal which removes a GroupContext extension that is present in the
+`required_capabilities` list is invalid. Adding a required capability that is
+not supported by all group members is already forbidden by {{!RFC9420}}.
+
+
+# Diff Formats
```

Comment:
This section needs to specify the structure of the extensions as well, so that it's clear how the diffs are applied.  My guess is that what you want is something like:

```
opaque Opaque<V>;

struct {
    Opaque key;
    Opaque value;
} KeyValuePair;


Value VectorExtension<V>;
KeyValuePair KeyValueExtension<V>;
```